src/dmd/expressionsem.d

```diff
@@ -10225,6 +10225,24 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             return setError();
         }
 
+        if (t1.ty == Tclass && t2.ty == Tclass)
```

Comment:
I suspect this is handled too early here. What about operator overloads, e.g. opBinary and opBinaryRight?

IMO lowerings should happen after all possible semantic errors have been reported to the user, so the user doesn't see cryptic messages referring to the lowered expressions in druntime (unless the implementation there is broken).

Reply:
> I suspect this is handled too early here. IMO lowerings should happen after all possible semantic errors have been reported to the user, so the user doesn't see cryptic messages referring to the lowered expressions in druntime (unless the implementation there is broken).

Where, specifically, do you recommend this be done?  How does one know that semantic is finished and it's the proper time for lowering?

> What about operator overloads, e.g. opBinary and opBinaryRight?

Can you provide a test case?

Reply:
Sorry for the delay.

> Where, specifically, do you recommend this be done?

I would have expected it to be close to where the array lowering is happening, i.e. after operator overload.

> How does one know that semantic is finished and it's the proper time for lowering?

Yeah, hard to tell due to possibly deferred analysis. I imagine/dream that no lowerings should happen in the first 3 semantic passes (so user messages never get polluted with compiler generated identifiers and the AST can be mapped back to the source correctly). An additional pass would lower the code to library constructs and continue semantic analysis on the result, but should never generate messages if the library code isn't screwed up.

> > What about operator overloads, e.g. opBinary and opBinaryRight?
> Can you provide a test case?

I was wrong about opBinary that cannot replace opCmp, but here's an example that gets lowered too early:
```
import core.stdc.stdio;

class A
{
	int a = 3;
	
	override int opCmp(Object o)
	{
		auto b = cast(B)o;
		if (!b)
			return super.opCmp(o);
		printf("comparing a=%d with b=%d\n", this.a, b.b);
		return this.a - b.b;
	}
}

class B
{
	int b = 4;
}

void main()
{
	A a = new A;
	B b = new B;
	assert(b > a);
}
```
I don't think it is possible for the library code to detect "rewrite 2" in https://dlang.org/spec/operatoroverloading.html#compare

Reply:
> Yeah, hard to tell due to possibly deferred analysis.

Doesn't the semantic analysis end here:

https://github.com/dlang/dmd/blob/40d85cf5ab5094230ab37368273b331c85211a96/src/dmd/mars.d#L619

Or possibly here:

https://github.com/dlang/dmd/blob/40d85cf5ab5094230ab37368273b331c85211a96/src/dmd/mars.d#L653